src/CloudFoundry.WinDEA/StagingInstance.cs

```diff
@@ -265,34 +267,226 @@ public void GetBuildpack(StagingStartMessageRequest message, string gitPath, str
                     throw new Exception("Buildpack does not support this application.");
                 }
             }
+            else if (message.Properties.BuildpackKey != null)
+            {
+                Logger.Info("Staging task {0}: Using admin buildpack {1}", this.Properties.TaskId, message.Properties.BuildpackKey);
+
+                TryCleanupUnrackedAdminBuildpacks(message.AdminBuildpacks, adminBuildpackDir);
+                var ab = message.AdminBuildpacks.First(i => i.Key == message.Properties.BuildpackKey);
+
+                var buildpackMutex = new Mutex(false, "admin-buildpack-" + ab.Key);
+                try
+                {
+                    buildpackMutex.WaitOne();
```

Comment:
I think it is safe without a timeout. It is just used to guard the downloading and copying to the container from each other and skip the cleanup if it is in use.